frontend/src/extensions.ts

```diff
@@ -6,44 +6,152 @@
 import { get as store_get } from "svelte/store";
 
 import { getUrlPath, urlFor } from "./helpers";
+import { fetch } from "./lib/fetch";
 import { log_error } from "./log";
 import { extensions } from "./stores";
 
+class ExtensionApi {
+  extension_name: string;
+
+  constructor(extension_name: string) {
+    this.extension_name = extension_name;
+  }
+
+  request(
+    endpoint: string,
+    method: string,
+    params: Record<string, string | number> | undefined,
+    body: object | undefined,
+    output: "json" | "string" | "raw" = "json",
+  ) {
+    const url = urlFor(
+      `extension/${this.extension_name}/${endpoint}`,
+      params,
+      false,
+    );
+    let opts = {};
+    if (body) {
+      opts =
+        body instanceof FormData
+          ? { body }
+          : {
+              headers: { "Content-Type": "application/json" },
+              body: JSON.stringify(body),
+            };
+    }
+    const request = fetch(url, { method, ...opts });
+    if (output === "json") {
+      return request.then((res) => res.json());
+    }
+    if (output === "string") {
+      return request.then((res) => res.text());
+    }
+    return request;
+  }
+
+  get(
+    endpoint: string,
+    params: Record<string, string | number>,
+    output: "json" | "string" | "raw" = "json",
+  ) {
+    return this.request(endpoint, "GET", params, undefined, output);
+  }
+
+  put(
+    endpoint: string,
+    body: object | undefined,
+    output: "json" | "string" | "raw" = "json",
+  ) {
+    return this.request(endpoint, "PUT", undefined, body, output);
+  }
+
+  post(
+    endpoint: string,
+    body: object | undefined,
+    output: "json" | "string" | "raw" = "json",
+  ) {
+    return this.request(endpoint, "POST", undefined, body, output);
+  }
+
+  delete(
+    endpoint: string,
+    body: object | undefined,
+    output: "json" | "string" | "raw" = "json",
+  ) {
+    return this.request(endpoint, "DELETE", undefined, body, output);
+  }
+}
+
+export interface ExtensionContext {
+  api: ExtensionApi;
+  getExt: (name: string) => Promise<ExtensionData>;
+}
+
 /**
  * The Javascript code of a Fava extension should export an object of this type.
  */
 export interface ExtensionModule {
   /** Initialise this Javascript module / run some code on the initial load. */
-  init?: () => void;
+  init?: (c: ExtensionContext) => void;
   /** Run some code after any Fava page has loaded. */
-  onPageLoad?: () => void;
+  onPageLoad?: (c: ExtensionContext) => void;
   /** Run some code after the page for this extension has loaded. */
-  onExtensionPageLoad?: () => void;
+  onExtensionPageLoad?: (c: ExtensionContext) => void;
+}
```

Comment:
I also ended up adding this:
```
init(c) {
    this.api = c.api;
},
```
to a bunch of the extensions I've been testing/prototyping with, since the inline event handler extension that I moved that inline loading code to sets up event handlers for other extensions that need a context, but aren't being called from within an "extension entrypoint", so it doesn't have access to those contexts. So I'm just storing it manually on the extension module so I can reference it from anywhere. It was also a little easier than having to pass around a context within the extension module.
What if we used the ExtensionModule type to specify the interface that extension JS code should export, but specify that Fava's extension initialization code will "construct" the module into a subtype of ExtensionModule that includes things like a context variable (or just api variable directly). That subtype would roughly be like what ExtensionData is in this change, but if we went this route I think that could be made into an interface instead of a class.

Reply:
I've refactored the ExtensionData class to be an internal implementation detail and not be exposed to the extension. I do not think that our code should be doing any mutation on the ExtensionModule or constructing some subtype. If this continues to be a pain point, I'd suggest a different solution: allow the init extension handler to also register the onPageLoad and onExtensionPageLoad handlers. That way, you can construct your extension class/function/module/whatever in the init function and have the context easily available.

Reply:
Ok no problem, it's not much of a pain point anyways, and I think setting `this.api = c.api;` in init() is a good enough workaround/hack for me for when I don't feel like using the context properly.

I also noticed you removed the 'output' parameter from the get, post, etc. functions. Is that because you want to treat that as a lower level parameter so if you want non-json you should just use request() directly?